vector/sidecar_server.py

```diff
@@ -1,43 +1,72 @@
 #!/usr/bin/env python3
-"""Launch vector in a subprocess and handle web signaling for the sidecar."""
+"""Launch vector in a subprocess and listen for POST /quitquitquit. Also keep an eye on
+/var/log/sidecar-log-consumer/heartbeat, which may exist and will have a unix timestamp
+representing the last heartbeat from the airflow container.
+"""
+import time
 import os
 import subprocess
 from http import server
+from pathlib import Path
 
 ppid = os.getppid()
 
 proc = subprocess.Popen("/usr/local/bin/vector", shell=True)
 
 
-class ExitHandler(server.BaseHTTPRequestHandler):
+def quit_proc():
+    try:
+        print("Terminating.", flush=True)
+        proc.terminate()
+        proc.wait(timeout=60)
+    except subprocess.TimeoutExpired:
+        print("Termination timed out. Killing.", flush=True)
+        proc.kill()
+
+
+class MessageHandler(server.BaseHTTPRequestHandler):
+    """Listen for a POST to /quitquitquit to signal that the sidecar should exit. Also keep
+    an eye on the heartbeat file to see if the primary container has been brutally killed.
+    """
+
     def do_POST(self):
         if self.path != "/quitquitquit":
             return
-        print("Exiting.")
+
         self.send_response(200)
         self.end_headers()
 
-        try:
-            outs, errs = proc.communicate(timeout=5)
-        except subprocess.TimeoutExpired:
-            proc.kill()
-            outs, errs = proc.communicate()
-        if outs:
-            print(f"{outs=}")
-        if errs:
-            print(f"{errs=}")
+        quit_proc()
         raise SystemExit()
 
 
-class ThisServer(server.HTTPServer):
+class MessageServer(server.HTTPServer):
+    heartbeat_file = Path("/var/log/sidecar-log-consumer/heartbeat")
+    heartbeat_max_age = 180  # seconds
+    last_timestamp = None
+
     def service_actions(self):
         proc.poll()
         if proc.returncode is not None:
             raise SystemExit(proc.returncode)
 
+        if self.heartbeat_file.exists():
+            # Sometimes the file contents are empty due to a race condition, so we only update
+            # last_timestamp if the file contents can be converted to a float.
+            if last_timestamp := float(self.heartbeat_file.read_text()):
+                self.last_timestamp = last_timestamp
+            age = time.time() - self.last_timestamp
+            if age > self.heartbeat_max_age / 2:
+                print(
+                    f"WARNING: Heartbeat has not been sent for {age:0.1f} seconds",
+                    flush=True,
+                )
```

Comment:
This section is really chatty. It produces one message every 0.5 seconds. I'm not sure it's even needed.